File: emulinker/src/main/java/org/emulinker/config/RuntimeFlags.java
package org.emulinker.config;

import static com.google.common.base.Preconditions.checkArgument;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import java.nio.charset.Charset;
import java.time.Duration;
import java.util.List;
import javax.inject.Singleton;
import org.apache.commons.configuration.Configuration;

/** Configuration flags that are set at startup and do not change until the job is restarted. */
@AutoValue
@Singleton
public abstract class RuntimeFlags {

  public abstract boolean allowMultipleConnections();

  public abstract boolean allowSinglePlayer();

  public abstract boolean metricsEnabled();

  public abstract boolean touchEmulinker();

  public abstract boolean touchKaillera();

  public abstract Charset charset();

  public abstract Duration metricsLoggingFrequency();

  public abstract ImmutableList<String> connectionTypes();

  public abstract int coreThreadPoolSize();

  public abstract int chatFloodTime();

  public abstract int createGameFloodTime();

  public abstract int gameAutoFireSensitivity();

  public abstract int gameBufferSize();

  public abstract int gameDesynchTimeouts();

  public abstract int gameTimeoutMillis();

  public abstract int idleTimeout();

  public abstract int keepAliveTimeout();

  public abstract int maxChatLength();

  public abstract int maxClientNameLength();

  public abstract int maxGameChatLength();

  public abstract int maxGameNameLength();

  public abstract int maxGames();

  public abstract int maxPing();

  public abstract int maxQuitMessageLength();

  public abstract int maxUserNameLength();

  public abstract int maxUsers();

  public abstract String serverAddress();

  public abstract String serverLocation();

  public abstract String serverName();

  public abstract String serverWebsite();

  public abstract int v086BufferSize();

  public static Builder builder() {
    return new AutoValue_RuntimeFlags.Builder();
  }

  public static RuntimeFlags loadFromApacheConfiguration(Configuration config) {
    return builder()
        .setAllowMultipleConnections(config.getBoolean("server.allowMultipleConnections"))
        .setAllowSinglePlayer(config.getBoolean("server.allowSinglePlayer", true))
        .setCharset(Charset.forName(config.getString("emulinker.charset")))
        .setChatFloodTime(config.getInt("server.chatFloodTime"))
        .setConnectionTypes(config.getList("server.allowedConnectionTypes"))
        // TODO(nue): Experiment with Runtime.getRuntime().availableProcessors().
        .setCoreThreadPoolSize(config.getInt("server.coreThreadpoolSize", 5))
        .setCreateGameFloodTime(config.getInt("server.createGameFloodTime"))
        .setMetricsEnabled(config.getBoolean("metrics.enabled", false))
        .setMetricsLoggingFrequency(
            Duration.ofSeconds(config.getInt("metrics.loggingFrequencySeconds", 30)))
        .setGameAutoFireSensitivity(config.getInt("game.defaultAutoFireSensitivity"))
        .setGameBufferSize(config.getInt("game.bufferSize"))
        .setGameDesynchTimeouts(config.getInt("game.desynchTimeouts"))
        .setGameTimeoutMillis(config.getInt("game.timeoutMillis"))
        .setIdleTimeout(config.getInt("server.idleTimeout"))
        .setKeepAliveTimeout(config.getInt("server.keepAliveTimeout"))
        .setMaxChatLength(config.getInt("server.maxChatLength"))
        .setMaxClientNameLength(config.getInt("server.maxClientNameLength"))
        .setMaxGameChatLength(config.getInt("server.maxGameChatLength"))
        .setMaxGameNameLength(config.getInt("server.maxGameNameLength"))
        .setMaxGames(config.getInt("server.maxGames"))
        .setMaxPing(config.getInt("server.maxPing"))
        .setMaxQuitMessageLength(config.getInt("server.maxQuitMessageLength"))
        .setMaxUserNameLength(config.getInt("server.maxUserNameLength"))
        .setMaxUsers(config.getInt("server.maxUsers"))
        .setServerAddress(config.getString("masterList.serverConnectAddress", ""))
        .setServerLocation(config.getString("masterList.serverLocation", "Unknown"))
        .setServerName(config.getString("masterList.serverName", "Emulinker Server"))
        .setServerWebsite(config.getString("masterList.serverWebsite", ""))
        .setTouchEmulinker(config.getBoolean("masterList.touchEmulinker", false))
        .setTouchKaillera(config.getBoolean("masterList.touchKaillera", false))
        .setV086BufferSize(config.getInt("controllers.v086.bufferSize", 4096))
        .build();
  }

  private void validateFlags() {
    // Note: this used to be max 30, but for some reason we had 31 set as the default in the config.
    // Setting this to max 31 so we don't break existing users.
    // TODO(nue): Just remove this restriction as it seems unhelpful?
    checkArgument(maxUserNameLength() <= 31, "server.maxUserNameLength must be <= 31");
    checkArgument(maxGameNameLength() <= 127, "server.maxGameNameLength must be <= 127");
    checkArgument(maxClientNameLength() <= 127, "server.maxClientNameLength must be <= 127");
    checkArgument(maxPing() > 0, "server.maxPing can not be <= 0");
    checkArgument(maxPing() <= 1000, "server.maxPing can not be > 1000");
    checkArgument(
        keepAliveTimeout() > 0, "server.keepAliveTimeout must be > 0 (190 is recommended)");
    checkArgument(gameBufferSize() > 0, "game.bufferSize can not be <= 0");
    checkArgument(gameTimeoutMillis() > 0, "game.timeoutMillis can not be <= 0");
    checkArgument(
        gameAutoFireSensitivity() >= 0 && gameAutoFireSensitivity() <= 5,
        "game.defaultAutoFireSensitivity must be 0-5");

    for (String s : connectionTypes()) {
      try {
        int ct = Integer.parseInt(s);
        checkArgument(ct >= 1 && ct <= 6, "Invalid connectionType: " + s);
      } catch (NumberFormatException e) {
        throw new IllegalStateException("Invalid connectionType: " + s, e);
      }
    }
  }

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract Builder setAllowMultipleConnections(boolean allowMultipleConnections);

    public abstract Builder setAllowSinglePlayer(boolean allowSinglePlayer);

    public abstract Builder setCharset(Charset charset);

    public abstract Builder setChatFloodTime(int chatFloodTime);

    public abstract Builder setConnectionTypes(List<String> connectionTypes);

    public abstract Builder setCoreThreadPoolSize(int coreThreadPoolSize);

    public abstract Builder setCreateGameFloodTime(int createGameFloodTime);

    public abstract Builder setMetricsEnabled(boolean enableMetrics);

    public abstract Builder setMetricsLoggingFrequency(Duration metricsLoggingFrequency);

    public abstract Builder setGameAutoFireSensitivity(int gameAutoFireSensitivity);

    public abstract Builder setGameBufferSize(int gameBufferSize);

    public abstract Builder setGameDesynchTimeouts(int gameDesynchTimeouts);

    public abstract Builder setGameTimeoutMillis(int gameTimeoutMillis);

    public abstract Builder setIdleTimeout(int idleTimeout);

    public abstract Builder setKeepAliveTimeout(int keepAliveTimeout);

    public abstract Builder setMaxChatLength(int maxChatLength);

    public abstract Builder setMaxClientNameLength(int maxClientNameLength);

    public abstract Builder setMaxGameChatLength(int maxGameChatLength);

    public abstract Builder setMaxGameNameLength(int maxGameNameLength);

    public abstract Builder setMaxGames(int maxGames);

    public abstract Builder setMaxPing(int maxPing);

    public abstract Builder setMaxQuitMessageLength(int maxQuitMessageLength);

    public abstract Builder setMaxUserNameLength(int maxUserNameLength);

    public abstract Builder setMaxUsers(int maxUsers);

    public abstract Builder setTouchEmulinker(boolean touchEmulinker);

    public abstract Builder setTouchKaillera(boolean touchKaillera);

    public abstract Builder setServerAddress(String serverAddress);

    public abstract Builder setServerLocation(String serverLocation);

    public abstract Builder setServerName(String serverName);

    public abstract Builder setServerWebsite(String serverWebsite);

    public abstract Builder setV086BufferSize(int v086BufferSize);

    protected abstract RuntimeFlags innerBuild();

    public final RuntimeFlags build() {
      RuntimeFlags flags = innerBuild();
      flags.validateFlags();
      return flags;
    }
  }
}

File: emulinker/src/main/java/org/emulinker/config/RuntimeFlags.kt
package org.emulinker.config

import java.nio.charset.Charset
import javax.inject.Singleton
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import org.apache.commons.configuration.Configuration

/** Configuration flags that are set at startup and do not change until the job is restarted. */
@Singleton
data class RuntimeFlags(
    val allowMultipleConnections: Boolean,
    val allowSinglePlayer: Boolean,
    val charset: Charset,
    val chatFloodTime: Int,
    val connectControllerBufferSize: Int,
    val connectionTypes: List<String>,
    val coreThreadPoolSize: Int,
    val createGameFloodTime: Int,
    val gameAutoFireSensitivity: Int,
    val gameBufferSize: Int,
    val gameDesynchTimeouts: Int,
    val gameTimeout: Duration,
    val idleTimeout: Duration,
    val improvedLagstatEnabled: Boolean,
    val keepAliveTimeout: Duration,
    val maxChatLength: Int,
    val maxClientNameLength: Int,
    val maxGameChatLength: Int,
    val maxGameNameLength: Int,
    val maxGames: Int,
    val maxPing: Int,
    val maxQuitMessageLength: Int,
    val maxUserNameLength: Int,
    val maxUsers: Int,
    val metricsEnabled: Boolean,
    val metricsLoggingFrequency: Duration,
    val requestTimeout: Duration,
    val serverAddress: String,
    val serverLocation: String,
    val serverName: String,
    val serverWebsite: String,
    val touchEmulinker: Boolean,
    val touchKaillera: Boolean,
    val twitterBroadcastDelay: Duration,
    val twitterEnabled: Boolean,
    val twitterOAuthAccessToken: String,
    val twitterOAuthAccessTokenSecret: String,
    val twitterOAuthConsumerKey: String,
    val twitterOAuthConsumerSecret: String,
    val twitterPreventBroadcastNameSuffixes: List<String>,
    val v086BufferSize: Int,
) {

  init {
    // Note: this used to be max 30, but for some reason we had 31 set as the default in the config.
    // Setting this to max 31 so we don't break existing users.
    // TODO(nue): Just remove this restriction as it seems unhelpful?
    require(maxUserNameLength <= 31) { "server.maxUserNameLength must be <= 31" }
    require(maxGameNameLength <= 127) { "server.maxGameNameLength must be <= 127" }
    require(maxClientNameLength <= 127) { "server.maxClientNameLength must be <= 127" }
    require(maxPing > 0) { "server.maxPing can not be <= 0" }
    require(maxPing <= 1000) { "server.maxPing can not be > 1000" }
    require(keepAliveTimeout.isPositive()) {
      "server.keepAliveTimeout must be > 0 (190 is recommended)"
    }

    require(gameBufferSize > 0) { "game.bufferSize can not be <= 0" }
    require(gameTimeout.isPositive()) { "game.timeoutMillis can not be <= 0" }
    require(gameAutoFireSensitivity in 0..5) { "game.defaultAutoFireSensitivity must be 0-5" }
    for (s in connectionTypes) {
      try {
        val ct = s.toInt()
        require(ct in 1..6) { "Invalid connectionType: $s" }
      } catch (e: NumberFormatException) {
        throw IllegalStateException("Invalid connectionType: $s", e)
      }
    }
  }

  companion object {

    fun loadFromApacheConfiguration(config: Configuration): RuntimeFlags {
      @Suppress("UNCHECKED_CAST") // TODO(nue): Replace commons-configurations.
      return RuntimeFlags(
          allowMultipleConnections = config.getBoolean("server.allowMultipleConnections"),
          allowSinglePlayer = config.getBoolean("server.allowSinglePlayer", true),
          charset = Charset.forName(config.getString("emulinker.charset")),
          chatFloodTime = config.getInt("server.chatFloodTime"),
          connectionTypes = config.getList("server.allowedConnectionTypes") as List<String>,
          coreThreadPoolSize = config.getInt("server.coreThreadpoolSize", 5),
          createGameFloodTime = config.getInt("server.createGameFloodTime"),
          gameAutoFireSensitivity = config.getInt("game.defaultAutoFireSensitivity"),
          gameBufferSize = config.getInt("game.bufferSize"),
          gameDesynchTimeouts = config.getInt("game.desynchTimeouts"),
          gameTimeout = config.getInt("game.timeoutMillis").milliseconds,
          idleTimeout = config.getInt("server.idleTimeout").seconds,
          improvedLagstatEnabled = config.getBoolean("server.improvedLagstatEnabled", true),
          keepAliveTimeout = config.getInt("server.keepAliveTimeout").seconds,
          maxChatLength = config.getInt("server.maxChatLength"),
          maxClientNameLength = config.getInt("server.maxClientNameLength"),
          maxGameChatLength = config.getInt("server.maxGameChatLength"),
          maxGameNameLength = config.getInt("server.maxGameNameLength"),
          maxGames = config.getInt("server.maxGames"),
          maxPing = config.getInt("server.maxPing"),
          maxQuitMessageLength = config.getInt("server.maxQuitMessageLength"),
          maxUserNameLength = config.getInt("server.maxUserNameLength"),
          maxUsers = config.getInt("server.maxUsers"),
          metricsEnabled = config.getBoolean("metrics.enabled", false),
          metricsLoggingFrequency = config.getInt("metrics.loggingFrequencySeconds", 30).seconds,
          requestTimeout =
              config.getInt(
                      "server.requestTimeoutMilliseconds", 5.seconds.inWholeMilliseconds.toInt())
                  .milliseconds,
          serverAddress = config.getString("masterList.serverConnectAddress", ""),
          serverLocation = config.getString("masterList.serverLocation", "Unknown"),
          serverName = config.getString("masterList.serverName", "Emulinker Server"),
          serverWebsite = config.getString("masterList.serverWebsite", ""),
          touchEmulinker = config.getBoolean("masterList.touchEmulinker", false),
          touchKaillera = config.getBoolean("masterList.touchKaillera", false),
          twitterBroadcastDelay = config.getInt("twitter.broadcastDelaySeconds", 15).seconds,
          twitterEnabled = config.getBoolean("twitter.enabled", false),
          twitterOAuthAccessToken = config.getString("twitter.auth.oAuthAccessToken", ""),
          twitterOAuthAccessTokenSecret =
              config.getString("twitter.auth.oAuthAccessTokenSecret", ""),
          twitterOAuthConsumerKey = config.getString("twitter.auth.oAuthConsumerKey", ""),
          twitterOAuthConsumerSecret =
              config.getString(
                  "twitter.auth.oAuthConsumerSecret", ""), // TODO(nue): Read these from a file
          twitterPreventBroadcastNameSuffixes = listOf("待", "街", "町", "再起", "侍ち"),
          v086BufferSize = config.getInt("controllers.v086.bufferSize", 4096),
          connectControllerBufferSize = config.getInt("controllers.connect.bufferSize", 4096),
      )
      // ImmutableList.copyOf(config.getString("twitter.preventBroadcastNameSuffixes",
      // "").split(",")))
    }
  }
}
